File: api/app/infrastructure/external/file_storage/tos_file_storage.py
from typing import Callable
from app.domain.repositories.uow import IUnitOfWork
from typing import BinaryIO
from typing import Tuple
from fastapi.concurrency import run_in_threadpool
from datetime import datetime
import uuid
from fastapi import UploadFile
from app.infrastructure.storage.tos import Tos
import logging
import os
from app.domain.external.file_storage import FileStorage
from app.domain.models.file import File

logger = logging.getLogger(__name__)


class TosFileStorage(FileStorage):
    def __init__(
        self, bucket: str, tos: Tos, uow_factory: Callable[[], IUnitOfWork]
    ) -> None:
        self.bucket = bucket
        self.tos = tos
        self.uow_factory = uow_factory
        self.uow: IUnitOfWork = uow_factory()

    async def upload_file(self, uploadFile: UploadFile) -> File:
        try:
            file_id = str(uuid.uuid4())
            _, file_extension = os.path.splitext(uploadFile.filename)
            if not file_extension:
                file_extension = ""
            date_path = datetime.now().strftime("%Y/%m/%d")
            tos_key = f"{date_path}/{file_id}{file_extension}"

            await run_in_threadpool(
                self.tos._client.put_object,
                bucket=self.bucket,
                content=uploadFile.file,
                key=tos_key,
            )
            logger.info("文件上传成功")

            file = File(
                id=file_id,
                filename=uploadFile.filename,
                key=tos_key,
                extension=file_extension,
                mime_type=uploadFile.content_type or "",
                size=uploadFile.size,
            )
            async with self.uow:
                await self.uow.file.save(file)

            return file
        except Exception as e:
            logger.error(f"上传文件失败: {e}")
            raise ValueError("上传失败")

    async def download_file(self, file_id: str) -> Tuple[BinaryIO, File]:
        try:
            async with self.uow:
                file = await self.uow.file.get_file_by_id(file_id)
                if not file:
                    raise ValueError(f"该文件不存在")

            response = await run_in_threadpool(
                self.tos._client.get_object,
                bucket=self.bucket,
                key=file.key,
            )

            return response.content, file
        except Exception as e:
            logger.error(f"下载文件失败")
            raise ValueError("下载失败")

File: api-node/src/infrastructure/file-storage/tos-file-storage.ts
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { getPrisma } from '../database/prisma';
import { getSettings } from '../../config';
import { FileRecord } from '../../domain/models';
import { getTosClient } from '../storage/tos-client';
import { logger } from '../../utils/logger';
import { FileStorage } from './file-storage';

export class TosFileStorage implements FileStorage {
  private readonly bucket: string;

  constructor(bucket = getSettings().tosBucket) {
    this.bucket = bucket;
  }

  async uploadFile(file: Express.Multer.File): Promise<FileRecord> {
    const fileId = uuidv4();
    const ext = path.extname(file.originalname);
    const now = new Date();
    const datePath = `${now.getFullYear()}/${String(now.getMonth() + 1).padStart(2, '0')}/${String(now.getDate()).padStart(2, '0')}`;
    const tosKey = `${datePath}/${fileId}${ext}`;

    try {
      await getTosClient().putObject({
        bucket: this.bucket,
        key: tosKey,
        body: file.buffer,
        contentType: file.mimetype || undefined,
      });

      const record: FileRecord = {
        id: fileId,
        filename: file.originalname,
        filepath: '',
        key: tosKey,
        extension: ext.replace('.', ''),
        mime_type: file.mimetype,
        size: file.size,
      };

      await getPrisma().file.create({
        data: {
          id: record.id,
          filename: record.filename,
          filepath: record.filepath,
          key: record.key,
          extension: record.extension,
          mimeType: record.mime_type,
          size: record.size,
        },
      });

      logger.info(`TOS upload success: ${tosKey}`);
      return record;
    } catch (e) {
      logger.error(`TOS upload failed: ${(e as Error).message}`);
      throw new Error('上传失败');
    }
  }

  async getFileInfo(fileId: string): Promise<FileRecord> {
    const file = await getPrisma().file.findUnique({ where: { id: fileId } });
    if (!file) throw new Error('找不到此文件');
    return {
      id: file.id,
      filename: file.filename,
      filepath: file.filepath,
      key: file.key,
      extension: file.extension,
      mime_type: file.mimeType,
      size: file.size,
    };
  }

  async downloadFile(fileId: string): Promise<{ buffer: Buffer; file: FileRecord }> {
    const file = await this.getFileInfo(fileId);
    try {
      const { data } = await getTosClient().getObjectV2({
        bucket: this.bucket,
        key: file.key,
        dataType: 'buffer',
      });
      const buffer = data.content as Buffer;
      return { buffer, file };
    } catch (e) {
      logger.error(`TOS download failed: ${(e as Error).message}`);
      throw new Error('下载失败');
    }
  }
}
